examples/resnet-concurrent.cpp

```diff
@@ -0,0 +1,164 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "glow/Base/Image.h"
+#include "glow/ExecutionEngine/ExecutionEngine.h"
+#include "glow/Graph/Graph.h"
+#include "glow/Importer/Caffe2ModelLoader.h"
+#include "glow/Runtime/RuntimeTypes.h"
+
+// We have not written the piece in the HostManager that initializes and creates
+// DeviceManagers for individual devices, so for now just reach into lib/
+#include "Backends/CPU/CPUDeviceManager.h"
+
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/FileSystem.h"
+
+#include <future>
+
+using namespace glow;
+using namespace glow::runtime;
+
+namespace {
+llvm::cl::OptionCategory category("resnet-concurrent Options");
+static llvm::cl::opt<std::string>
+    inputDirectory(llvm::cl::desc("input directory"),
+                   llvm::cl::init("../tests/images/imagenet/"),
+                   llvm::cl::Positional, llvm::cl::cat(category));
+llvm::cl::opt<unsigned> numDevices("numDevices",
+                                   llvm::cl::desc("Number of Devices to use"),
+                                   llvm::cl::init(7), llvm::cl::value_desc("N"),
+                                   llvm::cl::cat(category));
+} // namespace
+
+/// Run ResNet concurrently on a fixed number of CPU Devices
+int main(int argc, char **argv) {
+  llvm::cl::ParseCommandLineOptions(
+      argc, argv, "Run ResNet concurrently on a fixed number of CPU devices");
+
+  llvm::outs() << "Loading files from " << inputDirectory << "\n";
+
+  llvm::outs() << "Initializing " << numDevices << " CPU Devices.\n";
+  std::vector<std::unique_ptr<CPUDeviceManager>> devices;
+  for (int i = 0; i < numDevices; ++i) {
+    devices.emplace_back(std::make_unique<CPUDeviceManager>());
+    devices[i]->init();
+  }
+
+  Module module;
+  Function *F = module.createFunction("resnet50");
+  TypeRef inputType(module.uniqueType(ElemKind::FloatTy, {1, 3, 224, 224}));
+
+  llvm::outs() << "Loading resnet50 model.\n";
+
+  const char inputName[] = "gpu_0/data";
+  Caffe2ModelLoader loader("resnet50/predict_net.pb", "resnet50/init_net.pb",
+                           {inputName}, {inputType}, *F);
+  Placeholder *input =
+      llvm::cast<Placeholder>(cantFail(loader.getNodeValueByName(inputName)));
+  Placeholder *output = cantFail(loader.getSingleOutput());
+  auto *backend = createBackend(BackendKind::CPU);
```

Comment:
why not through ExecutionEngine?

Reply:
I mean the whole point is that we're using the DeviceManager, which the ExecutionEngine doesn't use. I could still do the compile through the EE then pull out the Module and the compiledFunction. Doesn't really feel clearer to me but I'm closer to the implementation.

Reply:
got it, makes sense. do we need any includes, dependencies on EE?

Reply:
Right now we do because `updateInputPlaceholders` is defined in the ExecutionEngine. I think we could move that func somewhere more generic, maybe to Context?